plugin/core/collections.py

```diff
@@ -0,0 +1,115 @@
+"""
+Module with additional collections.
+"""
+from .typing import Optional, Dict, Any
+
+
+class DottedDict:
+
+    __slots__ = ('_d')
+
+    def __init__(self, d: Optional[Dict[str, Any]] = None) -> None:
+        """
+        Construct a DottedDict, optionally from an existing dictionary.
+
+        :param      d:    An existing dictionary.
+        """
+        self._d = {}  # type: Dict[str, Any]
+        if d is not None:
+            self.update(d)
+
+    def get(self, path: Optional[str] = None) -> Any:
+        """
+        Get a value from the dictionary.
+
+        :param      path:  The path, e.g. foo.bar.baz, or None.
+
+        :returns:   The value stored at the path, or None if it doesn't exist.
+                    Note that this cannot distinguish between None values and
+                    paths that don't exist. If the path is None, returns the
+                    entire dictionary.
+        """
+        if path is None:
+            return self._d
+        current = self._d  # type: Any
+        keys = path.split('.')
+        for key in keys:
+            if isinstance(current, dict):
+                current = current.get(key)
+            else:
+                return None
+        return current
+
+    def set(self, path: str, value: Any) -> None:
+        """
+        Set a value in the dictionary.
+
+        :param      path:   The path, e.g. foo.bar.baz
+        :param      value:  The value
+        """
+        current = self._d
+        keys = path.split('.')
+        for i in range(0, len(keys) - 1):
```

Comment:
Isn't this the same while not looking like it's gonna make the brain explode:
```py
for i in range(len(keys)):
```
?

Or even shorter and cleaner:
```py
for i, key in enumerate(keys):
```

But then it doesn't look like you need index here so maybe just:
```py
for key in keys:
```
?

Reply:
The last chunk of the path gets special treatment (`current[keys[-1]] = value`). For the earlier chunks: we want to turn the thing the chunk points to into a dict if it isn't a dict already:

```
            if not isinstance(next_current, dict):
                next_current = {}
                current[key] = next_current
```
so that we can traverse deeper into `self._d`.

Reply:
Right. Sorry.